rust/kernel/virtio.rs

```diff
@@ -0,0 +1,187 @@
+    types::PointerWrapper, Result, ThisModule,
+};
+
+/// A registration of a virtio driver.
+pub type Registration<T> = driver::Registration<Adapter<T>>;
+
+/// Id of an Virtio device.
+#[derive(Clone, Copy)]
+pub struct DeviceId {
+    /// Device id.
+    pub device: u32,
+
+    /// Vendor id of the virtio device.
+    pub vendor: u32,
+}
+
+// SAFETY: `ZERO` is all zeroed-out and `to_rawid` stores in `virtio_id::data`.
+unsafe impl const driver::RawDeviceId for DeviceId {
+    type RawType = bindings::virtio_device_id;
+    const ZERO: Self::RawType = bindings::virtio_device_id {
+        device: 0,
+        vendor: 0,
+    };
+
+    fn to_rawid(&self, _offset: isize) -> Self::RawType {
+        bindings::virtio_device_id {
+            device: self.device,
+            vendor: self.vendor,
+        }
+    }
+}
+
+/// An adapter for the registration of virtio drivers.
+pub struct Adapter<T: Driver>(T);
+
+impl<T: Driver> driver::DriverOps for Adapter<T> {
+    type RegType = bindings::virtio_driver;
+
+    unsafe fn register(
+        reg: *mut bindings::virtio_driver,
+        name: &'static CStr,
+        module: &'static ThisModule,
+    ) -> Result {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` is non-null and valid.
+        let vdrv = unsafe { &mut *reg };
+
+        vdrv.driver.name = name.as_char_ptr();
+        vdrv.driver.owner = module.0;
+        vdrv.probe = Some(Self::probe_callback);
+        vdrv.remove = Some(Self::remove_callback);
+        if let Some(t) = T::ID_TABLE {
+            vdrv.id_table = t.as_ref();
+        }
+        // SAFETY:
+        //   - `vdrv` lives at least until the call to `register_virtio_driver()` returns.
+        //   - `name` pointer has static lifetime.
+        //   - `module.0` lives at least as long as the module.
+        //   - `probe()` and `remove()` are static functions.
+        //   - `id_table` is either a raw pointer with static lifetime,
+        //      as guaranteed by the [`driver::IdTable`] type, or null.
+        to_result(unsafe { bindings::register_virtio_driver(reg) })
+    }
+
+    unsafe fn unregister(reg: *mut bindings::virtio_driver) {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` was passed (and updated) by a previous successful call to
+        // `register_virtio_driver`.
+        unsafe { bindings::unregister_virtio_driver(reg) };
+    }
+}
+
+impl<T: Driver> Adapter<T> {
+    extern "C" fn probe_callback(vdev: *mut bindings::virtio_device) -> core::ffi::c_int {
+        from_kernel_result! {
+            // SAFETY: `vdev` is valid by the contract with the C code. `dev` is alive only for the
+            // duration of this call, so it is guaranteed to remain alive for the lifetime of
+            // `vdev`.
+            let mut dev = unsafe { Device::from_ptr(vdev) };
+            let data = T::probe(&mut dev)?;
```

Comment:
`of` identifiers are allowed to carry some data with them. If the match happened through one of them, ideally you'll pass this data to the `probe` function. See `get_id_info` in `platform.rs` for an example of this. (Perhaps we should move this function to the `driver` module so all drivers that support `of` can use it.)

Reply:
The virtio drivers use `id_table` to match the driver with the device, which does not contain any `data` field. The `probe` function in the virtio `Driver` is more like the `probe` function in the amba `Driver`.
It's reasonable to move `get_id_info` into the `driver` module. I can open a new PR about it.